### PR TITLE
Improving smartjoin for non-strings (#3080)

### LINKED ISSUE #3080: should keywords be treated as string (for smartjoin)
**Nikola Version:** 8.0.0beta1

I get the following:
```
    callable_(context, *args, **kwargs)
  File "/home/tbm/homepage/cache/.mako.tmp/base.tmpl.py", line 73, in render_body
    context['self'].extra_head(**pageargs)
  File "/home/tbm/homepage/cache/.mako.tmp/post.tmpl.py", line 164, in render_extra_head
    __M_writer(filters.html_escape(str(smartjoin(', ', post.meta('keywords')))))
  File "/home/tbm/.local/lib/python3.5/site-packages/nikola/utils.py", line 1884, in smartjoin
    return join_char.join(string_or_iterable)
TypeError: sequence item 4: expected str instance, int found
```

caused by:
```
keywords: [Debian, QNAP, TS-409, Orion, 5281]
```

5281 is interpreted as an integer instead of a string.  I wonder if all keywords should be converted to strings.

### PR DESCRIPTION
**Pull request by felixfontein**
Fixes #3080.

### CODE REVIEW

File: nikola/utils.py [hunk 1]
```diff
@@ -1870,17 +1870,19 @@ def sort_posts(posts, *keys):
...
     >>> smartjoin('; ', 'foo, bar')
     'foo, bar'
     >>> smartjoin('; ', ['foo', 'bar'])
     'foo; bar'
+    >>> smartjoin(' to ', ['count', 42])
+    'count to 42'
     """
-    if isinstance(string_or_iterable, (unicode_str, bytes_str)):
-        return string_or_iterable
+    if hasattr(string_or_iterable, '__iter__'):
```
[the remark this answers is not in the document]
Comment (felixfontein) [reply nikola/utils.py]: Ooops. That was stupid :)

[the remark this answers is not in the document]
Comment (felixfontein) [reply nikola/utils.py]: Fixed that...

File: nikola/utils.py [hunk 2]
```diff
@@ -1870,17 +1870,21 @@ def sort_posts(posts, *keys):
...
     >>> smartjoin('; ', 'foo, bar')
     'foo, bar'
     >>> smartjoin('; ', ['foo', 'bar'])
     'foo; bar'
+    >>> smartjoin(' to ', ['count', 42])
+    'count to 42'
     """
     if isinstance(string_or_iterable, (unicode_str, bytes_str)):
         return string_or_iterable
+    elif hasattr(string_or_iterable, '__iter__'):
```
Comment (Kwpolska) [inline nikola/utils.py]: `isinstance(string_or_iterable, collections.Iterable)` would look nicer IMO

Comment (felixfontein) [reply nikola/utils.py]: Is this actually equivalent?

Comment (Kwpolska) [reply nikola/utils.py]: yes

Comment (felixfontein) [reply nikola/utils.py]: Good to know. I changed it.

**Pull request merged**